llvm/lib/Target/MSP430/MSP430EpilogueOptimization.cpp

```diff
@@ -0,0 +1,146 @@
+///
+/// Technically, sometimes those epilogues can be emitted just as a 2-byte
+/// relative JMP instead of absolute 4-byte BR. Current implementation never
+/// makes code larger, at least.
+///
+/// See MSP430 Embedded Application Binary Interface, part 3.8
+/// "__mspabi_func_epilog Helper Functions" for details.
+//===----------------------------------------------------------------------===//
+
+#include "MSP430.h"
+#include "MSP430TargetMachine.h"
+
+#include "llvm/CodeGen/MachineFunctionPass.h"
+#include "llvm/CodeGen/MachineInstrBuilder.h"
+
+using namespace llvm;
+
+namespace {
+
+class MSP430EpilogueOptimization : public MachineFunctionPass {
+public:
+  MSP430EpilogueOptimization() : MachineFunctionPass(ID) {}
+
+  bool runOnMachineFunction(MachineFunction &MF) override;
+
+private:
+  static char ID;
+  const MSP430InstrInfo *TII;
+
+  void replaceTailWithBrToLibCall(MachineBasicBlock &MBB,
+                                   MachineInstr &FirstInstrToReplace,
+                                   const char *LibCallName);
+
+  bool tryReplacingEpilogue(MachineBasicBlock &MBB);
+};
+
+char MSP430EpilogueOptimization::ID = 0;
+
+} // end anonymous namespace
+
+/// \brief LibCall names arranged so that `N`-th element restores N registers
+///
+/// MSP430 EABI defines standard epilogue helper functions like these:
+/// \code
+/// __mspabi_func_epilog_7: RET R4
+/// __mspabi_func_epilog_6: RET R5
+/// __mspabi_func_epilog_5: RET R6
+/// __mspabi_func_epilog_4: RET R7
+/// __mspabi_func_epilog_3: RET R8
+/// __mspabi_func_epilog_2: RET R9
+/// __mspabi_func_epilog_1: RET R10
+///                         RET
+/// \endcode
+/// @note These helper functions do not exist on MSP430X
+static const char *EpilogueLibCallNames[] = {
+    nullptr,
+    "__mspabi_func_epilog_1",
+    "__mspabi_func_epilog_2",
+    "__mspabi_func_epilog_3",
+    "__mspabi_func_epilog_4",
+    "__mspabi_func_epilog_5",
+    "__mspabi_func_epilog_6",
+    "__mspabi_func_epilog_7",
+};
+
+/// Registers to be restored, from the last `POP` instruction to the first
+static const unsigned RestoredRegisters[] = {
+    MSP430::R10, MSP430::R9, MSP430::R8, MSP430::R7,
+    MSP430::R6,  MSP430::R5, MSP430::R4,
+};
+
+void MSP430EpilogueOptimization::replaceTailWithBrToLibCall(
+    MachineBasicBlock &MBB, MachineInstr &FirstInstrToReplace,
+    const char *LibCallName) {
+
+  const DebugLoc &DL = FirstInstrToReplace.getDebugLoc();
+  BuildMI(MBB, FirstInstrToReplace, DL, TII->get(MSP430::Bi))
+      .addExternalSymbol(LibCallName);
+
+  for (MachineInstr *MI = &FirstInstrToReplace; MI != nullptr;) {
+    MachineInstr *InstrToRemove = MI;
+    MI = MI->getNextNode();
+    InstrToRemove->eraseFromParent();
+  }
+}
+
+bool MSP430EpilogueOptimization::tryReplacingEpilogue(MachineBasicBlock &MBB) {
+  MachineInstr *LastMatchedPopInstr = nullptr;
+  unsigned MatchedRegisters = 0;
+
+  MachineBasicBlock::iterator Terminator = MBB.getFirstTerminator();
```

Comment:
Ok, after thinking about this a bit more I believe we have an issue here. It's perfectly ok for a basic block to contain multiple terminators. For example, it could be a conditional branch and then return, so taking first terminator would not work here. I think you could craft a testcase with a simple do {} while() loop and return afterwards.

Reply:
Agree, I have tested with something like
```ll
define i16 @f() noinline optnone {
entry:
  br label %do.cond

do.cond:
  br i1 1, label %do.cond, label %do.end

do.end:
  ret i16 42
}
```
being converted to
```
$ /path/to/llc -stop-after=legalizer < while.ll
// ...
body:             |
  bb.0.entry:
    successors: %bb.1(0x80000000)
  
    JMP %bb.1
  
  bb.1.do.cond:
    successors: %bb.1(0x40000000), %bb.2(0x40000000)
  
    $r12 = MOV16rc 0
    CMP16rc killed $r12, 1, implicit-def $sr
    JCC %bb.1, 1, implicit $sr
    JMP %bb.2
  
  bb.2.do.end:
    $r12 = MOV16ri 42
    RET implicit $r12
```
Not yet sure how to trick code generator into emitting conditional jump and return inside a single MBB, but this is definitely worth fixing.

Reply:
No need to trick. Use llc --print-after to dump MIR. You could edit it as you wish and run codegen on it :)

Probably in your case 'ret void' will work ;) Though... no additional code is allowed between the terminators (unless the code is predicated), so we cannot restore regs there. So, maybe the approach will still work, yes, though is a bit fragile... I'd scan from the end of BB for the sake of clarity.

Reply:
So, maybe it is worth leaving it as is, just convert the last for-loop in the `replaceTailWithBrToLibCall`
```cpp
for (MachineInstr *MI = &FirstInstrToReplace; MI != nullptr;) { ...
```
to remove till `FirstInstrToReplace`, *inclusive*, instead of removing till the end. And add an explanation. This way, the modification would produce identical behavior and leave MBB not less well-formed than it was before :)

Reply:
I have pushed a fixed version but without an explanation yet. Is there some documentation on "no additional code is allowed between the terminators" and other such things? Looks like it is less documented than the LLVM IR.